cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
   friend array<Y> const_array_cast(const array<U>&);
 
   public:
-    explicit array(std::int64_t element_count)
-        : data_(new T[element_count]), // TODO: Use custom oneDAL allocator
-          size_(element_count) {}
-
-    explicit array(const detail::shared<T>& data, std::int64_t size)
-        : data_(data),
-          size_(size) {}
+    array()
+        : data_owned_ptr_(nullptr),
+          size_(0),
+          capacity_(0) {}
+
+    explicit array(std::int64_t element_count, T element = {})
+        : data_owned_ptr_(new T[element_count], // TODO: Use custom oneDAL allocator
+                [](T* obj){ delete[] obj; }),
+          data_(data_owned_ptr_.get()),
+          size_(element_count),
+          capacity_(element_count) {
+        for (std::int64_t i = 0; i < size_; i++) {
+            (*this)[i] = element;
+        }
```

Comment:
I suggest to not always fill the array with the constant value. In other words, split this constructor into two, 
- Constructor that simply allocates the memory;
- Constructor that allocates the memory and fills it with the constant value;

Reply:
The other option is to introduce static functions like `full`/`zeros`/`ones` (see NumPy). It would give a hint to the user how to interpret this "constructor".

Reply:
I like the option of static functions. Moreover, I am thinking about the case of DPC++, where these construction operations will be dependent on the context. I think it is better to design them as free functions, @rlnx what is your opinion?
What's about constructor which simply allocates memory - afaik, such operations are prohibited by SDL requirements.

Reply:
SDL -- there is an exception for performance libraries, we are not obliged to always initialize the memory once it's allocated.